pkg/data/generic.go

```diff
@@ -79,167 +87,175 @@ func (s *GenericSync) RunContext(ctx context.Context) error {
 	if s.createError != nil {
 		return s.createError
 	}
-	s.loop(ctx.Done())
+
+	store, queue := s.setup(ctx)
+	defer queue.ShutDown()
+
+	go s.loop(ctx, store, queue)
+	<-ctx.Done()
 	return nil
 }
 
-func (s *GenericSync) loop(quit <-chan struct{}) {
+// setup the store and queue for this GenericSync instance
+func (s *GenericSync) setup(ctx context.Context) (cache.Store, workqueue.Interface) {
 
-	defer func() {
-		logrus.Infof("Sync for %v finished. Exiting.", s.ns)
-	}()
-
-	resource := s.client.Resource(schema.GroupVersionResource{
+	baseResource := s.client.Resource(schema.GroupVersionResource{
 		Group:    s.ns.Group,
 		Version:  s.ns.Version,
 		Resource: s.ns.Resource,
 	})
+	var resource dynamic.ResourceInterface = baseResource
+	if s.ns.Namespaced {
+		resource = baseResource.Namespace(metav1.NamespaceAll)
+	}
 
-	delay := backoffMin
+	queue := workqueue.New()
+	store, controller := cache.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+				return resource.List(ctx, options)
+			},
+			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+				return resource.Watch(ctx, options)
+			},
+		},
+		&unstructured.Unstructured{},
+		0,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				if key, err := cache.MetaNamespaceKeyFunc(obj); err == nil {
+					var _ string = key // Tiny sanity check. The code depends on queue keys being strings.
```

Comment:
This is https://pkg.go.dev/k8s.io/client-go/tools/cache#MetaNamespaceKeyFunc, right? I don't think we need this. What does this give us? 💭

Reply:
Yes, that's the function. It actually all comes down from the decision of using a `workqueue`. The previous code did all the work from inside the `ResourceHandler`, but issue #134 is about separating the kubernetes control loop from the OPA control loop. So I decided to have a `workqueue` to decouple the `ResourceHandler` from the rest of the controller logic.

With a `workqueue` you usually push the cache key of the object instead of the object itself, and then the controller uses the key to retrieve the latest version of the object from the `store` of the informer. So we need the `MetaNamespaceKeyFunc` to get the key of the object in the store.

Reply:
Sorry I should have been more explicit. What I meant was more trivial, when that function returns `(string, error)`, what do we get from `var _ string = key`?

Reply:
Ah sorry. There is a comment in the function documentation,

> TODO: replace key-as-string with a key-as-struct so that this packing/unpacking won't be necessary. 

I added the type check in case we upgrade the dependency in the future, and they have actually replaced the string with some struct. All the code that handles keys uses `interface{}`, so the compiler might happily compile our code without noticing.

Reply:
💡 good point. I wasn't aware of this. 👍